File: 43173/frontend/src/api/community.ts
import request, { PageData, PageParams } from '@/utils/request'

export interface Comment {
  id: number
  user_id: number
  work_id?: number
  album_id?: number
  playlist_id?: number
  parent_id?: number
  content: string
  like_count: number
  reply_count: number
  is_pinned: boolean
  user?: {
    id: number
    username: string
    nickname: string
    avatar: string
  }
  replies?: Comment[]
  created_at: string
}

export interface Playlist {
  id: number
  user_id: number
  name: string
  description?: string
  cover_url: string
  is_public: boolean
  play_count: number
  like_count: number
  follow_count: number
  work_count: number
  user?: {
    id: number
    username: string
    nickname: string
    avatar: string
  }
  works?: any[]
  created_at: string
  updated_at: string
}

export interface Notification {
  id: number
  user_id: number
  type: string
  title: string
  content: string
  data?: string
  is_read: boolean
  created_at: string
}

export interface PlayRecord {
  id: number
  user_id: number
  work_id: number
  artist_id: number
  duration: number
  ip: string
  device: string
  source: string
  work?: any
  created_at: string
}

export const communityApi = {
  follow: (followingId: number) => request.post('/follow', { following_id: followingId }),
  unfollow: (followingId: number) => request.post('/unfollow', { following_id: followingId }),
  isFollowing: (userId: number) => request.get<{ is_following: boolean }>(`/follow/is-following/${userId}`),
  getFollowers: (userId: number, params: PageParams) => request.get<PageData<any>>(`/users/${userId}/followers`, params),
  getFollowings: (userId: number, params: PageParams) => request.get<PageData<any>>(`/users/${userId}/followings`, params),
  getFollowingArtists: (params: PageParams) => request.get<PageData<any>>('/follow/artists', params),
  getFollowingUsers: (params: PageParams) => request.get<PageData<any>>('/follow/users', params),
  getMyFollowers: (params: PageParams) => request.get<PageData<any>>('/follow/followers', params),
  getFollowerCount: (userId: number) => request.get<{ count: number }>(`/users/${userId}/follower-count`),
  getFollowingCount: (userId: number) => request.get<{ count: number }>(`/users/${userId}/following-count`),
  
  createComment: (data: { work_id?: number; album_id?: number; playlist_id?: number; parent_id?: number; content: string }) => 
    request.post<Comment>('/comments', data),
  deleteComment: (id: number) => request.delete(`/comments/${id}`),
  getComments: (params: { work_id?: number; album_id?: number; playlist_id?: number; page?: number; page_size?: number }) => 
    request.get<PageData<Comment>>('/comments', params),
  
  createPlaylist: (data: { name: string; description?: string; cover_url?: string; is_public?: boolean }) => 
    request.post<Playlist>('/playlists', data),
  updatePlaylist: (id: number, data: { name?: string; description?: string; cover_url?: string; is_public?: boolean }) => 
    request.put(`/playlists/${id}`, data),
  deletePlaylist: (id: number) => request.delete(`/playlists/${id}`),
  getPlaylistById: (id: number) => request.get<Playlist>(`/playlists/${id}`),
  getMyPlaylists: () => request.get<Playlist[]>('/my-playlists'),
  listPlaylists: (params: { user_id?: number; keyword?: string; page?: number; page_size?: number }) => 
    request.get<PageData<Playlist>>('/playlists', params),
  addWorkToPlaylist: (playlistId: number, workId: number) => 
    request.post('/playlists/works', { playlist_id: playlistId, work_id: workId }),
  removeWorkFromPlaylist: (playlistId: number, workId: number) => 
    request.delete('/playlists/works', { playlist_id: playlistId, work_id: workId }),
  
  getNotifications: (params: PageParams) => request.get<PageData<Notification>>('/notifications', params),
  markNotificationAsRead: (id: number) => request.put(`/notifications/${id}/read`),
  markAllNotificationsAsRead: () => request.put('/notifications/read-all'),
  getUnreadNotificationCount: () => request.get<{ count: number }>('/notifications/unread-count'),
  
  getPlayRecords: (params: PageParams) => request.get<PageData<PlayRecord>>('/play-records', params)
}

File: 43173/backend/internal/model/community.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type Follow struct {
	ID          uint           `json:"id" gorm:"primaryKey"`
	FollowerID  uint           `json:"follower_id" gorm:"index;not null"`
	FollowingID uint           `json:"following_id" gorm:"index;not null"`
	CreatedAt   time.Time      `json:"created_at"`
	DeletedAt   gorm.DeletedAt `json:"-" gorm:"index"`
}

func (Follow) TableName() string {
	return "follows"
}

type Comment struct {
	ID         uint           `json:"id" gorm:"primaryKey"`
	UserID     uint           `json:"user_id" gorm:"index;not null"`
	WorkID     uint           `json:"work_id" gorm:"index"`
	AlbumID    *uint          `json:"album_id" gorm:"index"`
	PlaylistID *uint          `json:"playlist_id" gorm:"index"`
	ParentID   *uint          `json:"parent_id" gorm:"index"`
	Content    string         `json:"content" gorm:"type:text;not null"`
	LikeCount  int64          `json:"like_count" gorm:"default:0"`
	ReplyCount int64          `json:"reply_count" gorm:"default:0"`
	IsPinned   bool           `json:"is_pinned" gorm:"default:false"`
	Status     int            `json:"status" gorm:"default:1"`
	User       *User          `json:"user,omitempty" gorm:"foreignKey:UserID"`
	Replies    []Comment      `json:"replies,omitempty" gorm:"foreignKey:ParentID"`
	CreatedAt  time.Time      `json:"created_at"`
	UpdatedAt  time.Time      `json:"updated_at"`
	DeletedAt  gorm.DeletedAt `json:"-" gorm:"index"`
}

func (Comment) TableName() string {
	return "comments"
}

type Playlist struct {
	ID           uint           `json:"id" gorm:"primaryKey"`
	UserID       uint           `json:"user_id" gorm:"index;not null"`
	Title        string         `json:"title" gorm:"size:200;not null"`
	Description  string         `json:"description" gorm:"type:text"`
	CoverURL     string         `json:"cover_url" gorm:"size:255"`
	IsPublic     bool           `json:"is_public" gorm:"default:true"`
	PlayCount    int64          `json:"play_count" gorm:"default:0"`
	LikeCount    int64          `json:"like_count" gorm:"default:0"`
	FollowCount  int64          `json:"follow_count" gorm:"default:0"`
	WorkCount    int            `json:"work_count" gorm:"default:0"`
	User         *User          `json:"user,omitempty" gorm:"foreignKey:UserID"`
	Works        []Work         `json:"works,omitempty" gorm:"many2many:playlist_works"`
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
	DeletedAt    gorm.DeletedAt `json:"-" gorm:"index"`
}

func (Playlist) TableName() string {
	return "playlists"
}

type PlaylistWork struct {
	PlaylistID uint `gorm:"primaryKey"`
	WorkID     uint `gorm:"primaryKey"`
	AddedAt    time.Time `json:"added_at"`
}

func (PlaylistWork) TableName() string {
	return "playlist_works"
}

type Like struct {
	ID        uint           `json:"id" gorm:"primaryKey"`
	UserID    uint           `json:"user_id" gorm:"index;not null"`
	WorkID    *uint          `json:"work_id" gorm:"index"`
	AlbumID   *uint          `json:"album_id" gorm:"index"`
	CommentID *uint          `json:"comment_id" gorm:"index"`
	CreatedAt time.Time      `json:"created_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`
}

func (Like) TableName() string {
	return "likes"
}

type PlayRecord struct {
	ID        uint           `json:"id" gorm:"primaryKey"`
	UserID    uint           `json:"user_id" gorm:"index;not null"`
	WorkID    uint           `json:"work_id" gorm:"index;not null"`
	ArtistID  uint           `json:"artist_id" gorm:"index;not null"`
	Duration  int            `json:"duration"`
	IP        string         `json:"ip" gorm:"size:45"`
	Device    string         `json:"device" gorm:"size:100"`
	Source    string         `json:"source" gorm:"size:50"`
	CreatedAt time.Time      `json:"created_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`
}

func (PlayRecord) TableName() string {
	return "play_records"
}

type Share struct {
	ID         uint           `json:"id" gorm:"primaryKey"`
	UserID     uint           `json:"user_id" gorm:"index;not null"`
	WorkID     *uint          `json:"work_id" gorm:"index"`
	AlbumID    *uint          `json:"album_id" gorm:"index"`
	PlaylistID *uint          `json:"playlist_id" gorm:"index"`
	Platform   string         `json:"platform" gorm:"size:50"`
	IP         string         `json:"ip" gorm:"size:45"`
	CreatedAt  time.Time      `json:"created_at"`
	DeletedAt  gorm.DeletedAt `json:"-" gorm:"index"`
}

func (Share) TableName() string {
	return "shares"
}

type Notification struct {
	ID        uint           `json:"id" gorm:"primaryKey"`
	UserID    uint           `json:"user_id" gorm:"index;not null"`
	Type      string         `json:"type" gorm:"size:50"`
	Title     string         `json:"title" gorm:"size:200"`
	Content   string         `json:"content" gorm:"type:text"`
	Data      string         `json:"data" gorm:"type:text"`
	IsRead    bool           `json:"is_read" gorm:"default:false"`
	CreatedAt time.Time      `json:"created_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`
}

func (Notification) TableName() string {
	return "notifications"
}
